src/main/java/com/iota/iri/utils/thread/ThreadUtils.java

```diff
@@ -0,0 +1,139 @@
+package com.iota.iri.utils.thread;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.HashMap;
+
+/**
+ * This class contains a collection of methods that make the management of {@link Thread}s a little bit more convenient.
+ *
+ * Apart from offering method to spawn and terminate {@link Thread}s, it also offers methods to {@link #sleep(int)}
+ * while correctly maintaining the {@link Thread#isInterrupted()} flags.
+ */
+public class ThreadUtils {
+    /**
+     * Logger for this class allowing us to dump debug and status messages.
+     */
+    private static final Logger logger = LoggerFactory.getLogger(ThreadUtils.class);
+
+    /**
+     * Holds a map of {@link Thread}s that are managed by the {@link ThreadUtils}.
+     *
+     * This map is used to offer a thread safe way to deal with spawning and stopping {@link Thread}s.
+     *
+     * @see #spawnThread(Runnable, ThreadIdentifier)
+     * @see #stopThread(ThreadIdentifier)
+     */
+    private static final HashMap<Object, Thread> threads = new HashMap<>();
+
+    /**
+     * This method spawns a new {@link Thread} for the given {@link ThreadIdentifier}.
+     *
+     * This method is thread safe and makes sure there is exactly one "non-interrupted" {@link Thread} associated to the
+     * identifier.
+     *
+     * It can therefore be used as a convenient way to offer a {@code start()} method in the corresponding manager
+     * instance, without having to worry about starting multiple {@link Thread}s when calling the {@code start()} method
+     * multiple times.
+     *
+     * @param runnable method reference that is responsible for processing the thread (i.e. {@code this::myThreadMethod})
+     * @param threadIdentifier identifier object that is used to associate the {@link Thread} and synchronize the access
+     * @return the thread that got spawned by this method
+     */
+    public static Thread spawnThread(Runnable runnable, ThreadIdentifier threadIdentifier) {
+        if (threads.get(threadIdentifier) == null || threads.get(threadIdentifier).isInterrupted()) {
+            synchronized(threadIdentifier) {
+                if (threads.get(threadIdentifier) == null || threads.get(threadIdentifier).isInterrupted()) {
```

Comment:
I don't think that using a ConcurrentHashMap will be sufficient because you are also writing to the map after which also have to be synchronized. So you want to synchronize those 2 calls rather than just have the get synchronized

Reply:
As I spoke to you in private I prefer that you use the `SingleThreadExecutor` or `ScheduledSingleThreadExecutor` (which can replace this class), Instead of doing the global scope solution.
This is because:
1. global scope is evil
2. I don't think that an "infrastructure" `spawnThread` method should have a `synchronized` in it. Especially when it is easy enough to do a solution that doesn't necessitate this.

Sorry for not telling you about the `SingleThreadExecutor` before. When I said you can wrap the logic that creates thread I didn't think you would do this global scope solution to manage threads.